### PR TITLE
Added optional expression tags

### PR DESCRIPTION
**Pull request by jaredoconnell**
## Changes introduced with this PR

This PR adds two more ways of specifying inferred inputs/outputs:
- `!wait-optional` - Creates a completion dependency that still resolves if the referenced expression fails, but it will wait until either it resolves or it is marked unresolvable. This is the one intended for most use cases. For example, if you don't care if a step succeeds or crashes, `!wait-optional` is the best one for that use case. But if you want the output to only resolve when the step is disabled or succeeds, but not if it crashes, you must use `!ordisabled`.
- `!soft-optional` - Creates an optional dependency that does not effect resolution of the object. There are no order guarantees, so this field is only intended for things that really do not matter strongly for the output.

These expressions may only be used as fields properties within an object. The fields will be marked as optional.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: workflow/workflow_test.go [hunk 1]
```diff
@@ -2111,6 +2111,402 @@ outputs:
...
+    input:
+      wait_time_ms: !expr $.input.wait_2_ms
+    enabled: !expr $.input.wait_2_enabled
+outputs:
+  workflow-success:
+    required_output: !expr $.steps.wait_step_1.outputs.success
+    optional_output: !soft-optional $.steps.wait_step_2.outputs.success
+`
+
+func TestSoftOptionalFieldWorkflow(t *testing.T) {
```
Comment (mfleader) [inline workflow/workflow_test.go:2268]: Can you explain why this test needs to use a comparison of time delays to make assertions about the source code behavior, and if possible add a succinct version of that in the test's header comment?

**Review by mfleader — CHANGES_REQUESTED** [review]

File: workflow/workflow_test.go (hunk 1 above)
Comment (jaredoconnell) [reply workflow/workflow_test.go:2268]: My goal was to isolate the soft-optional dependency as much as possible. Soft-optional provides no guarantees for ordering, so the design of this test is to use a large enough time to make it all but guaranteed that the intended outputs will be included.

**Comment by jaredoconnell** [comment]
> Are there any edge cases that you know of that are intentionally left uncovered by these tests?

Theoretically a soft-optional dependency could be included in more test cases, including alongside ORs, but that feature is, by design, not as reliable, and I didn't want to make the tests too fragile.

File: workflow/workflow.go [hunk 2]
```diff
@@ -712,7 +714,9 @@ func (l *loopState) resolveExpressions(inputData any, dataModel any) (any, error
 			if err != nil {
 				return nil, fmt.Errorf("failed to resolve workflow map expressions (%w)", err)
 			}
-			result[key] = newValue
+			if newValue != nil { // In case it's an optional field.
```
[the remark this answers is not in the document]
Comment (jaredoconnell) [reply workflow/workflow.go:717]: I don't think we have enough context at that spot. Missing data will be detected later when the data is serialized or unserialized.

File: workflow/executor.go [hunk 3]
```diff
@@ -1006,6 +1008,56 @@ func (e *executor) prepareExprDependencies(
...
+	dag dgraph.DirectedGraph[*DAGItem],
+) error {
+	var dependencyType dgraph.DependencyType
+	if expr.WaitForCompletion {
+		dependencyType = dgraph.CompletionAndDependency
+	} else {
+		dependencyType = dgraph.OptionalDependency
+	}
+
+	// The way that this will work is the current node will depend on the grouped
```
[the remark this answers is not in the document]
Comment (jaredoconnell) [reply workflow/executor.go]: Updated.

File: workflow/executor.go [hunk 4]
```diff
@@ -1006,6 +1008,56 @@ func (e *executor) prepareExprDependencies(
...
+	// node to isolate the optional dependencies. The current node will depend on it
```
[the remark this answers is not in the document]
Comment (jaredoconnell) [reply workflow/executor.go]: I re-worded it.

File: workflow/yaml_test.go [hunk 5]
```diff
@@ -169,7 +163,42 @@ func TestBuildResultOrDisabledExpression_InvalidPattern(t *testing.T) {
...
+func TestBuildWaitOptionalExpr_InvalidExpr(t *testing.T) {
+	yamlInput := []byte(`!wait-optional ....`)
+	input := assert.NoErrorR[yaml.Node](t)(yaml.New().Parse(yamlInput))
+	_, err := buildOptionalExpression(input, make([]string, 0))
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "placed in invalid configuration")
+}
+
+func TestBuildWaitOptionalExpr_InvalidTag(t *testing.T) {
+	// It should never get there outside the unit test.
```
[the remark this answers is not in the document]
Comment (jaredoconnell) [reply workflow/yaml_test.go]: I updated the comment.

**Review by dbutenhof — APPROVED** [review]

**Review by mfleader — APPROVED** [review]
